src/defrag.c

```diff
@@ -1159,17 +1162,33 @@ void activeDefragCycle(void) {
                 db = NULL;
                 server.active_defrag_running = 0;
 
+                /* If the defragmentation hits and defragmented bytes are both
+                 * below 1%, which indicates low defragmentation efficiency.
+                 * Reduce the defragmentation speed by 10% for the next cycle. */
+                long long last_hits = server.stat_active_defrag_hits - start_hits;
+                long long last_misses = server.stat_active_defrag_misses - start_misses;
+                if (last_hits < (last_hits + last_misses) * 0.01 &&
+                    start_frag_bytes >= frag_bytes &&
```

Comment:
`start_frag_bytes >= frag_bytes`, i'm not sure about that.
it could fluctuate up and down slightly, and i don' think it's an indication of anything. IMHO the only reason look at bytes do that is when the server is actually nearly empty, is when the fragmentation ratio appears to be high, but it's not a lot of memory.